### ISSUE TITLE
Message pact fails after upgrade to 3.6.5

### ISSUE THREAD

**Issue by Bjoern-Hegstam** [DiUS/pact-jvm#874]
I was glad to see that metadata verification had been implemented in 3.6.5 and wanted to try it out, but after upgrading from 3.6.4 to 3.6.5 I ran into some issues. The short summary is that to me it seems like the handling of content-type and contents has changed.

Here's the fragment setup from my provider:
```java
@Pact(provider = PROVIDER, consumer = CONSUMER)
public MessagePact fragment_systemAccessCreated(MessagePactBuilder builder) {
    return builder
            .given("A system access is created")
            .expectsToReceive("SystemAccessCreated message")
            .withContent(
                    new PactDslJsonBody()
                            .stringValue("origin", "staff-access")
                            .uuid("correlationId")
                            .stringValue("type", "SystemAccessCreated")
                            .object("version")
                            .numberValue("major", 1)
                            .numberType("minor", 0)
                            .closeObject()
                            .object("payload")
                            .uuid("systemId", "3e850e12-8d81-4be2-a0cf-3080bd63af7b")
                            .uuid("userId")
                            .closeObject()
            )
            .toPact();
}
```

when using 3.6.4 the following json file is generated

```json
{
  "consumer": {
    "name": "operator-service"
  },
  "provider": {
    "name": "system-access-change-events"
  },
  "messages": [
    {
      "description": "SystemAccessCreated message",
      "metaData": {
        "Content-Type": "application/json; charset\u003dUTF-8"
      },
      "contents": {
        "payload": {
          "systemId": "3e850e12-8d81-4be2-a0cf-3080bd63af7b",
          "userId": "e2490de5-5bd3-43d5-b7c4-526e33f71304"
        },
        "origin": "staff-access",
        "correlationId": "e2490de5-5bd3-43d5-b7c4-526e33f71304",
        "type": "SystemAccessCreated",
        "version": {
          "major": 1,
          "minor": 0
        }
      },
      "providerStates": [
        {
          "name": "A system access is created"
        }
      ],
      "matchingRules": {
        "body": {
          "$.correlationId": {
            "matchers": [
              {
                "match": "regex",
                "regex": "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
              }
            ],
            "combine": "AND"
          },
          "$.version.minor": {
            "matchers": [
              {
                "match": "number"
              }
            ],
            "combine": "AND"
          },
          "$.payload.systemId": {
            "matchers": [
              {
                "match": "regex",
                "regex": "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
              }
            ],
            "combine": "AND"
          },
          "$.payload.userId": {
            "matchers": [
              {
                "match": "regex",
                "regex": "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
              }
            ],
            "combine": "AND"
          }
        }
      }
    }
  ],
  "metadata": {
    "pactSpecification": {
      "version": "3.0.0"
    },
    "pact-jvm": {
      "version": "3.6.4"
    }
  }
}
```

while 3.6.5 gives the following
```json
{
  "consumer": {
    "name": "operator-service"
  },
  "provider": {
    "name": "system-access-change-events"
  },
  "messages": [
    {
      "description": "SystemAccessCreated message",
      "metaData": {
        "Content-Type": "application/json; charset\u003dUTF-8"
      },
      "contents": "\"\\\"{\\\\\\\"payload\\\\\\\":{\\\\\\\"systemId\\\\\\\":\\\\\\\"3e850e12-8d81-4be2-a0cf-3080bd63af7b\\\\\\\",\\\\\\\"userId\\\\\\\":\\\\\\\"e2490de5-5bd3-43d5-b7c4-526e33f71304\\\\\\\"},\\\\\\\"origin\\\\\\\":\\\\\\\"staff-access\\\\\\\",\\\\\\\"correlationId\\\\\\\":\\\\\\\"e2490de5-5bd3-43d5-b7c4-526e33f71304\\\\\\\",\\\\\\\"type\\\\\\\":\\\\\\\"SystemAccessCreated\\\\\\\",\\\\\\\"version\\\\\\\":{\\\\\\\"major\\\\\\\":1,\\\\\\\"minor\\\\\\\":0}}\\\"\"",
      "providerStates": [
        {
          "name": "A system access is created"
        }
      ],
      "matchingRules": {
        "body": {
          "$.correlationId": {
            "matchers": [
              {
                "match": "regex",
                "regex": "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
              }
            ],
            "combine": "AND"
          },
          "$.version.minor": {
            "matchers": [
              {
                "match": "number"
              }
            ],
            "combine": "AND"
          },
          "$.payload.systemId": {
            "matchers": [
              {
                "match": "regex",
                "regex": "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
              }
            ],
            "combine": "AND"
          },
          "$.payload.userId": {
            "matchers": [
              {
                "match": "regex",
                "regex": "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
              }
            ],
            "combine": "AND"
          }
        }
      }
    }
  ],
  "metadata": {
    "pactSpecification": {
      "version": "3.0.0"
    },
    "pact-jvm": {
      "version": "3.6.5"
    }
  }
}
```

using the following verification in my consumer:
```java
@PactVerifyProvider("SystemAccessCreated message")
    public String verifySystemAccessCreatedMessage() {
        List<PublishRequest> publishedRequests = snsStub.getPublishedRequests(SnsSystemAccessChangeEventPublisher.SYSTEM_ACCESS_CHANGE_TOPIC_NAME);
        return publishedRequests.get(0).getMessage();
    }

    @TestTarget
    public final Target target = new AmqpTarget(singletonList(this.getClass().getPackage().getName() + ".*"));
```

works fine on 3.6.4 (provider and consumer), but after upgrading to 3.6.5 I get the following:
```

Verifying a pact between operator-service and system-access-change-events
  Given A system access is created
  SystemAccessCreated message
2019-04-24 11:28:09.439UTC INFO  [main] [AAH-Correlation-Id=fb3cc959-6b0a-4159-8d35-2760148a5210] o.r.Reflections - Reflections took 176 ms to scan 228 urls, producing 4 keys and 10 values 
    generates a message which
      has a matching body (FAILED)

Failures:

0) SystemAccessCreated message generates a message which has a matching body
      / -> mismatch



java.lang.AssertionError: 
0 - / -> [{mismatch=mismatch, diff=}]

	at au.com.dius.pact.provider.junit.target.BaseTarget.getAssertionError(BaseTarget.kt:71)
	at au.com.dius.pact.provider.junit.target.AmqpTarget.testInteraction(AmqpTarget.kt:50)
	at au.com.dius.pact.provider.junit.InteractionRunner$interactionBlock$statement$1.evaluate(InteractionRunner.kt:197)
	at au.com.dius.pact.provider.junit.RunStateChanges.evaluate(RunStateChanges.kt:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at au.com.dius.pact.provider.junit.InteractionRunner.run(InteractionRunner.kt:144)
	at au.com.dius.pact.provider.junit.PactRunner.runChild(PactRunner.kt:137)
	at au.com.dius.pact.provider.junit.PactRunner.runChild(PactRunner.kt:51)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

I did some debugging and noted that in `ResponseComparison#compareMessageBody` no body matcher is found due to `message.contentType` being `application/json; charset=UTF-8` in 3.6.5, while with 3.6.4 it is `application/json`.

To temporarily get around this I manually removed the charset from the content-type in the pact file and re-ran the verification, but then I encountered this error:
```

Verifying a pact between operator-service and system-access-change-events
  Given A system access is created
  SystemAccessCreated message
2019-04-24 11:32:27.494UTC INFO  [main] [AAH-Correlation-Id=33049b2b-30d2-4c23-ba3d-139eb32a4f7c] o.r.Reflections - Reflections took 172 ms to scan 228 urls, producing 4 keys and 10 values 
    generates a message which
      has a matching body (FAILED)

Failures:

0) SystemAccessCreated message generates a message which has a matching body
      $ -> Expected '"\"{\\\"payload\\\":{\\\"systemId\\\":\\\"3e850e12-8d81-4be2-a0cf-3080bd63af7b\\\",\\\"userId\\\":\\\"e2490de5-5bd3-43d5-b7c4-526e33f71304\\\"},\\\"origin\\\":\\\"staff-access\\\",\\\"correlationId\\\":\\\"e2490de5-5bd3-43d5-b7c4-526e33f71304\\\",\\\"type\\\":\\\"SystemAccessCreated\\\",\\\"version\\\":{\\\"major\\\":1,\\\"minor\\\":0}}\""' but received Map(payload -> Map(systemId -> f535038c-89c2-4754-821a-2fbbc314616d, userId -> e34963ab-bcb8-49af-9f39-d71d9ad7999f), version -> Map(major -> 1, minor -> 0), origin -> staff-access, correlationId -> 33049b2b-30d2-4c23-ba3d-139eb32a4f7c, type -> SystemAccessCreated)



java.lang.AssertionError: 
0 - $ -> [{mismatch=Expected '"\"{\\\"payload\\\":{\\\"systemId\\\":\\\"3e850e12-8d81-4be2-a0cf-3080bd63af7b\\\",\\\"userId\\\":\\\"e2490de5-5bd3-43d5-b7c4-526e33f71304\\\"},\\\"origin\\\":\\\"staff-access\\\",\\\"correlationId\\\":\\\"e2490de5-5bd3-43d5-b7c4-526e33f71304\\\",\\\"type\\\":\\\"SystemAccessCreated\\\",\\\"version\\\":{\\\"major\\\":1,\\\"minor\\\":0}}\""' but received Map(payload -> Map(systemId -> f535038c-89c2-4754-821a-2fbbc314616d, userId -> e34963ab-bcb8-49af-9f39-d71d9ad7999f), version -> Map(major -> 1, minor -> 0), origin -> staff-access, correlationId -> 33049b2b-30d2-4c23-ba3d-139eb32a4f7c, type -> SystemAccessCreated), diff=}]

	at au.com.dius.pact.provider.junit.target.BaseTarget.getAssertionError(BaseTarget.kt:71)
	at au.com.dius.pact.provider.junit.target.AmqpTarget.testInteraction(AmqpTarget.kt:50)
	at au.com.dius.pact.provider.junit.InteractionRunner$interactionBlock$statement$1.evaluate(InteractionRunner.kt:197)
	at au.com.dius.pact.provider.junit.RunStateChanges.evaluate(RunStateChanges.kt:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at au.com.dius.pact.provider.junit.InteractionRunner.run(InteractionRunner.kt:144)
	at au.com.dius.pact.provider.junit.PactRunner.runChild(PactRunner.kt:137)
	at au.com.dius.pact.provider.junit.PactRunner.runChild(PactRunner.kt:51)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

Any help would be much appreciated. Am I missing something or is it a bug?

**Comment by uglyog** [comment]
Oh, damn. It looks like the contents is getting encoded as a string, maybe double encoded.

**Comment by uglyog** [comment]
Released 3.6.6 with this fix
